packages/loopring_v3/contracts/impl/ProtocolRegistry.sol

```diff
@@ -0,0 +1,176 @@
+    }
+
+    mapping (address => Protocol) private protocols;
+    mapping (address => address) public exchangeToProtocolMap;
+    address[] public exchanges;
+
+    /// @dev The constructor must do NOTHING to support proxy.
+    constructor() public {}
+
+    /// @dev The default function will create an upgradabile exchange
+    ///      with on-chain data-availability.
+    function() external payable
+    {
+        forgeExchange(true, true);
+    }
+
+    function registerProtocol(
+        address protocol,
+        string  memory version
+        )
+        public
+        nonReentrant
+    {
+        require(protocol != address(0), "ZERO_ADDRESS");
+        require(bytes(version).length > 0, "INVALID_VERSION_LABEL");
+
+        ILoopring loopring = ILoopring(protocol);
+        address instance = loopring.deployExchange();
+
+        // Initialize the instance in a way that it's onwed by the protocol address itself.
+        // This is to prevent the default instance from being used by any entity.
+        // TODO(daniel): uncomment this
+        // uint id = loopring.initializeExchange(instance, false);
+        // require(id == uint(1), "DEFAULT_INSTANCE_MUST_HAVE_ID_1");
+
+        protocols[protocol] = Protocol(instance, version);
+    }
+
+    function getProtocol(
+        address protocol
+        )
+        public
+        view
+        returns (
+            address instance,
+            string  memory version
+        )
+    {
+        require(protocol != address(0), "INVALID_PROTOCOL");
+        Protocol storage p = protocols[protocol];
+        instance = p.instance;
+        version = p.version;
+        require(instance != address(0), "INVALID_INSTANCE");
+    }
+
+    function getProtocol()
+        external
+        view
+        returns (
+            address protocol,
+            address instance,
+            string  memory version
+        )
+    {
+        protocol = exchangeToProtocolMap[msg.sender];
+        Protocol storage p = protocols[protocol];
+        instance = p.instance;
+        version = p.version;
+        require(instance != address(0), "INVALID_INSTANCE");
+    }
+
+    function setDefaultProtocol(
+        address protocol
+        )
+        external
+        onlyOwner
+        nonReentrant
+    {
+        (address instance, ) = getProtocol(protocol);
+        require(instance != address(0), "INVALID_PROTOCOL");
+        defaultProtocol = protocol;
+    }
+
+    function getDefaultProtocol()
+        external
+        view
+        returns (
+            address protocol,
+            address instance,
+            string  memory version
+        )
+    {
+        require(defaultProtocol != address(0), "NO_DEFAULT");
+        protocol = defaultProtocol;
+        (instance, version) = getProtocol(protocol);
+    }
+
+    function forgeExchange(
+        address protocol,
+        bool    supportUpgradability,
+        bool    onchainDataAvailability
+        )
+        public
+        nonReentrant
```

Comment:
It's best to only use `nonReentrant` on external functions. `public` functions can also be called in the contract itself, so if a `public` function that is `nonReentrant` is called from an `external` functions that is also `nonReentrant` it will fail, which is not something you'd expect.

Reply:
ok.